internal/locate/region_cache.go

```diff
@@ -366,28 +408,36 @@ func (r *Region) checkRegionCacheTTL(ts int64) bool {
 	if _, err := util.EvalFailpoint("invalidateRegionCache"); err == nil {
 		r.invalidate(Other)
 	}
+	newTTL := int64(0)
 	for {
-		lastAccess := atomic.LoadInt64(&r.lastAccess)
-		if ts-lastAccess > regionCacheTTLSec {
+		ttl := atomic.LoadInt64(&r.ttl)
+		if ts > ttl {
 			return false
 		}
-		if r.checkSyncFlags(needExpireAfterTTL) || atomic.CompareAndSwapInt64(&r.lastAccess, lastAccess, ts) {
+		// skip updating TTL when:
+		// 1. the region has been marked as `needExpireAfterTTL`
+		// 2. the TTL is far away from ts (still within jitter time)
+		if r.checkSyncFlags(needExpireAfterTTL) || ttl > ts+regionCacheTTLSec {
+			return true
+		}
+		if newTTL == 0 {
+			newTTL = nextTTL(ts)
+		}
```

Comment:
```suggestion
		} else if newTTL <= ttl {
			return true
		}
```

Because `newTTL` is only calculated once
If the ttl is updated by another thread, it might be `newTLL < ttl`, we can skip updating ddl in this case.

```
// thread1
ttl1 := atomic.LoadInt64(&r.ttl)
newTTL1 = nextTTL(ts1)

// thread2
ttl2 := atomic.LoadInt64(&r.ttl)
newTTL2 = nextTTL(ts2) // newTTL2 > newTTL1
atomic.CompareAndSwapInt64(&r.ttl, ttl2, newTTL2) == true

// thread1
ttl1 := atomic.LoadInt64(&r.ttl) // second loop, ttl1 == newTTL2
atomic.CompareAndSwapInt64(&r.ttl, ttl1, newTTL1) // the ttl jumps back
```

Reply:
I think the constraint at L426 always holds. If ttl1(=newTTL2) > ts+regionCacheTTLSec, the function returns at L421, if not ttl1(=newTTL2) <= ts+regionCacheTTLSec <= ts+regionCacheTTLSec+randomJitter = newTTL1.

Reply:
Indeed.